### ISSUE TITLE
@property hides multiple type errors

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/mypy#11892]
**Bug Report**

This code defines `self._body` as `Optional[bytes]` and does not produce any type errors: https://github.com/aio-libs/aiohttp/blob/f1eb09b3756f4ed378a25f8a634f734952d2da35/aiohttp/web_response.py#L589
If I instead define the type on the class, like:
```
class Response(StreamResponse):
    _body: Optional[bytes]
```

Then I get a type error: `Incompatible types in assignment (expression has type "Optional[bytes]", base class "StreamResponse" defined the type as "None")  [assignment]`

**To Reproduce**

See next comment.

**Your Environment**

- Mypy version used: 0.930 (but, those type errors must have been there for months already)
- Mypy configuration options from `mypy.ini`: https://github.com/aio-libs/aiohttp/blob/master/.mypy.ini

**Comment by Dreamsorcerer** [comment]
OK, seems to be a problem with `@property`, managed a minimal reproducer. This produces no type errors, but there are actually 2 issues in this code:
```
from typing import Optional

class Foo:
    def __init__(self) -> None:
        self._a = None  # Foo._a is defined as None

class Bar(Foo):
    def __init__(self, a: Optional[bytes]) -> None:
        self.a = a  # setter says type should be bytes, but we are assigning Optional[bytes].

    @property
    def a(self) -> Optional[bytes]:
        return self._a

    @a.setter
    def a(self, thing: bytes) -> None:
        self._a: Optional[bytes] = thing  # We are defining _a as Optional[bytes], but superclass says it should be None.
```

**Comment by Dreamsorcerer** [comment]
It also causes incorrect type errors. If I try to assign to a property, it will give a type error if the type does not match the return type, when it should be checking the parameter type in the setter:
```
    @property
    def last_modified(self) -> Optional[datetime.datetime]: ...

    @last_modified.setter
    def last_modified(self, value: Optional[Union[int, float, datetime.datetime, str]]) -> None: ...
```

```
resp.last_modified = filepath.stat().st_mtime
```
```
error: Incompatible types in assignment (expression has type "float", variable has type "Optional[datetime]")  [assignment]
```

**Comment by hauntsaninja** [comment]
mypy doesn't do setters that take different types particularly well, see https://github.com/python/mypy/issues/3004 There's been some movement on that issue recently, so hopefully gets fixed soon

**Comment by AlexWaygood** [comment]
> This produces no type errors, but there are actually 2 issues in this code

On 0.941, mypy now [does report](https://mypy-play.net/?mypy=latest&python=3.10&flags=check-untyped-defs&gist=b678db0ce4e8eb53e99f630ce6dcd5ff) one of these errors (but still not the other one):

 ```python 
from typing import Optional

class Foo:
    def __init__(self) -> None:
        self._a = None  # Foo._a is defined as None

class Bar(Foo):
    def __init__(self, a: Optional[bytes]) -> None:
        self.a = a  # setter says type should be bytes, but we are assigning Optional[bytes].

    @property
    def a(self) -> Optional[bytes]:
        return self._a

    @a.setter
    def a(self, thing: bytes) -> None:
        self._a: Optional[bytes] = thing  # error: Incompatible types in assignment (expression has type "Optional[bytes]", base class "Foo" defined the type as "None")
```
